devnet/devnet.go

```diff
@@ -0,0 +1,175 @@
+package devnet
+
+import (
+	"context"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+func runCmdsWithLog(ctx context.Context, name string, commands [][]string) {
+	logFile, err := os.Create(name + ".log")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer logFile.Close()
+
+	for _, cmdArgs := range commands {
+		log.Printf("command for %s: %s", name, strings.Join(cmdArgs, " "))
+		cmd := exec.CommandContext(ctx, cmdArgs[0], cmdArgs[1:]...)
+		cmd.Stdout = logFile
+		cmd.Stderr = logFile
+		// If ctx.Err()!=nil, we cancelled the command via SIGINT.
+		if err := cmd.Run(); err != nil && ctx.Err() == nil {
+			log.Printf("%s; check %s for details", err, logFile.Name())
+			break
+		}
+	}
+}
+
+func runLotusDaemon(ctx context.Context, home string) {
+	cmds := [][]string{
+		{"lotus-seed", "genesis", "new", "localnet.json"},
+		{"lotus-seed", "pre-seal", "--sector-size=2048", "--num-sectors=10"},
+		{"lotus-seed", "genesis", "add-miner", "localnet.json",
+			filepath.Join(home, ".genesis-sectors", "pre-seal-t01000.json")},
+		{"lotus", "daemon", "--lotus-make-genesis=dev.gen",
+			"--genesis-template=localnet.json", "--bootstrap=false"},
+	}
+
+	runCmdsWithLog(ctx, "lotus-daemon", cmds)
+}
+
+func runLotusMiner(ctx context.Context, home string) {
+	cmds := [][]string{
+		{"lotus", "wait-api"}, // wait for lotus node to run
+
+		{"lotus", "wallet", "import",
+			filepath.Join(home, ".genesis-sectors", "pre-seal-t01000.key")},
+		{"lotus-miner", "init", "--genesis-miner", "--actor=t01000", "--sector-size=2048",
+			"--pre-sealed-sectors=" + filepath.Join(home, ".genesis-sectors"),
+			"--pre-sealed-metadata=" + filepath.Join(home, ".genesis-sectors", "pre-seal-t01000.json"),
+			"--nosync"},
+
+		// Starting in network version 13,
+		// pre-commits are batched by default,
+		// and commits are aggregated by default.
+		// This means deals could sit at StorageDealAwaitingPreCommit or
+		// StorageDealSealing for a while, going past our 10m test timeout.
+		{"sed", "-ri", "-e", "s/#BatchPreCommits\\ =\\ true/BatchPreCommits=false/",
+			filepath.Join(home, ".lotusminer", "config.toml")},
+
+		{"sed", "-ri", "-e", "s/#AggregateCommits\\ =\\ true/AggregateCommits=false/",
+			filepath.Join(home, ".lotusminer", "config.toml")},
+
+		{"lotus-miner", "run", "--nosync"},
+	}
```

Comment:
Would it make sense to put all these commands in a script file and just run the script file?